Flaky-test: TxnLogBufferedWriterTest.testMainProcess

TxnLogBufferedWriterTest.testMainProcess is flaky. It fails sporadically.

[example failure](https://github.com/apache/pulsar/runs/7449540085?check_suite_focus=true)

```
  Error:  Tests run: 25, Failures: 2, Errors: 0, Skipped: 13, Time elapsed: 24.634 s <<< FAILURE! - in org.apache.pulsar.transaction.coordinator.impl.TxnLogBufferedWriterTest
  Error:  testMainProcess(org.apache.pulsar.transaction.coordinator.impl.TxnLogBufferedWriterTest)  Time elapsed: 4.032 s  <<< FAILURE!
  org.awaitility.core.ConditionTimeoutException: Condition with org.apache.pulsar.transaction.coordinator.impl.TxnLogBufferedWriterTest was not fulfilled within 4 seconds.
  	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:167)
  	at org.awaitility.core.CallableCondition.await(CallableCondition.java:78)
  	at org.awaitility.core.CallableCondition.await(CallableCondition.java:26)
  	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:985)
  	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:954)
  	at org.apache.pulsar.transaction.coordinator.impl.TxnLogBufferedWriterTest.testMainProcess(TxnLogBufferedWriterTest.java:235)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
  	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at java.base/java.lang.Thread.run(Thread.java:833)
```

https://github.com/apache/pulsar/runs/7573727384?check_suite_focus=true

```
Error:  Tests run: 25, Failures: 2, Errors: 0, Skipped: 13, Time elapsed: 20.098 s <<< FAILURE! - in org.apache.pulsar.transaction.coordinator.impl.TxnLogBufferedWriterTest
  Error:  testMainProcess(org.apache.pulsar.transaction.coordinator.impl.TxnLogBufferedWriterTest)  Time elapsed: 0.116 s  <<< FAILURE!
  java.lang.AssertionError: expected [true] but found [false]
  	at org.testng.Assert.fail(Assert.java:99)
  	at org.testng.Assert.failNotEquals(Assert.java:1037)
  	at org.testng.Assert.assertTrue(Assert.java:45)
  	at org.testng.Assert.assertTrue(Assert.java:55)
  	at org.apache.pulsar.transaction.coordinator.impl.TxnLogBufferedWriterTest.testMainProcess(TxnLogBufferedWriterTest.java:250)
```

--- a/pulsar-transaction/coordinator/src/test/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriterTest.java
+++ b/pulsar-transaction/coordinator/src/test/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriterTest.java
@@ -86,28 +86,28 @@ public class TxnLogBufferedWriterTest extends MockedBookKeeperTestCase {
     public Object[][] mainProcessCasesProvider(){
         return new Object [][]{
                 // Normal.
-                {512, 1024 * 1024, 1, true, 2000, 2, 4, BookieErrorType.NO_ERROR, false},
+                {512, 1024 * 1024, 1, true, 2000, 4, BookieErrorType.NO_ERROR, false},
                 // The number of data writes is few.
-                {512, 1024 * 1024, 1, true, 100, 2, 4, BookieErrorType.NO_ERROR, false},
+                {512, 1024 * 1024, 1, true, 100, 4, BookieErrorType.NO_ERROR, false},
                 // The number of data writes is large.
-                {512, 1024 * 1024, 100, true, 20000, 5, 4, BookieErrorType.NO_ERROR, false},
+                {512, 1024 * 1024, 100, true, 20000, 4, BookieErrorType.NO_ERROR, false},
                 // Big data writes.
-                {512, 1024, 100, true, 3000, 6, 1024, BookieErrorType.NO_ERROR, false},
+                {512, 1024, 100, true, 3000, 1024, BookieErrorType.NO_ERROR, false},
                 // A batch has only one data
-                {1, 1024 * 1024, 100, true, 2000, 6, 4, BookieErrorType.NO_ERROR, false},
+                {1, 1024 * 1024, 100, true, 512, 4, BookieErrorType.NO_ERROR, false},
                 // A batch has only two data
-                {2, 1024 * 1024, 100, true, 1999, 4, 4, BookieErrorType.NO_ERROR, false},
+                {2, 1024 * 1024, 100, true, 1999, 4, BookieErrorType.NO_ERROR, false},
                 // Disabled the batch feature
-                {512, 1024 * 1024, 1, false, 2000, 4, 4, BookieErrorType.NO_ERROR, false},
+                {128, 1024 * 1024, 1, false, 500, 4, BookieErrorType.NO_ERROR, false},
                 // Bookie always error.
-                {512, 1024 * 1024, 1, true, 2000, 2, 4, BookieErrorType.ALWAYS_ERROR, false},
-                {512, 1024 * 1024, 1, false, 2000, 4, 4, BookieErrorType.ALWAYS_ERROR, false},
+                {128, 1024 * 1024, 1, true, 512, 4, BookieErrorType.ALWAYS_ERROR, false},
+                {128, 1024 * 1024, 1, false, 512, 4, BookieErrorType.ALWAYS_ERROR, false},
                 // Bookie sometimes error.
-                {512, 1024 * 1024, 1, true, 2000, 4, 4, BookieErrorType.SOMETIMES_ERROR, false},
-                {512, 1024 * 1024, 1, false, 2000, 4, 4, BookieErrorType.SOMETIMES_ERROR, false},
+                {128, 1024 * 1024, 1, true, 512, 4, BookieErrorType.SOMETIMES_ERROR, false},
+                {128, 1024 * 1024, 1, false, 512, 4, BookieErrorType.SOMETIMES_ERROR, false},
                 // TxnLogBufferedWriter sometimes close.
-                {512, 1024 * 1024, 1, true, 2000, 4, 4, BookieErrorType.NO_ERROR, true},
-                {512, 1024 * 1024, 1, false, 2000, 4, 4, BookieErrorType.NO_ERROR, true}
+                {128, 1024 * 1024, 1, true, 512, 4, BookieErrorType.NO_ERROR, true},
+                {128, 1024 * 1024, 1, false, 512, 4, BookieErrorType.NO_ERROR, true}
         };
     }
 
@@ -126,9 +126,9 @@ public class TxnLogBufferedWriterTest extends MockedBookKeeperTestCase {
      *      composite-ByteBuf.
      *   3. Even if executed "bkc.failAfter", also need to verify the data which has already written to bookie.
      */
-    @Test(dataProvider = "mainProcessCasesProvider")
+    @Test(dataProvider = "mainProcessCasesProvider", timeOut = 1000 * 10)
     public void testMainProcess(int batchedWriteMaxRecords, int batchedWriteMaxSize, int batchedWriteMaxDelayInMillis,
-                                boolean batchEnabled, final int writeCmdExecuteCount, int maxWaitSeconds,
+                                boolean batchEnabled, final int writeCmdExecuteCount,
                                 int eachDataBytesLen, BookieErrorType bookieErrorType,
                                 boolean closeBufferedWriter) throws Exception {
         // Assert args.
@@ -215,7 +215,7 @@ public class TxnLogBufferedWriterTest extends MockedBookKeeperTestCase {
             }
             if (closeBufferedWriter && bufferedWriteCloseAtIndex == i){
                 // Wait for any complete callback, avoid unstable.
-                Awaitility.await().atMost(2, TimeUnit.SECONDS).until(() -> anyFlushCompleted.get());
+                Awaitility.await().until(() -> anyFlushCompleted.get());
                 txnLogBufferedWriter.close();
             }
         }
@@ -231,8 +231,7 @@ public class TxnLogBufferedWriterTest extends MockedBookKeeperTestCase {
          *   not sort.
          */
         // Assert callback count.
-        Awaitility.await().atMost(maxWaitSeconds, TimeUnit.SECONDS)
-                .until(() -> contextArrayOfCallback.size() == writeCmdExecuteCount);
+        Awaitility.await().until(() -> contextArrayOfCallback.size() == writeCmdExecuteCount);
         // Assert callback param-context, verify that all callbacks are executed in strict order.
         if (closeBufferedWriter){
             Collections.sort(contextArrayOfCallback);

